File: games/billiards/server/rules/snooker.ts
import type { GameRandomV1 } from "@tabletop/game-sdk/server";
import type { SeatId } from "@tabletop/protocol";

import { GameRuleError } from "@tabletop/game-sdk/server";

import { tableSpecFor } from "../../shared/table.js";
import type { BilliardsDecidingBlackChoice } from "../../shared/actions.js";
import type { BilliardsBall, SnookerOn } from "../../shared/view.js";
import {
  addScore,
  ballById,
  endState,
  isSnookerColor,
  markCueBallInHand,
  newlyPottedBalls,
  otherSeat,
  requireCompetitivePlayers,
  SNOOKER_COLOR_VALUES,
  SNOOKER_COLORS_ASCENDING,
} from "./common.js";
import type {
  AdjudicatedBilliardsShot,
  BilliardsMatchState,
  BilliardsSimulationResult,
  ShotAdjudicationInput,
} from "../state.js";

function withPottedFlags(
  balls: readonly BilliardsBall[],
  pocketedIds: readonly string[],
  cuePotted: boolean,
): readonly BilliardsBall[] {
  const pocketed = new Set(pocketedIds);
  return balls.map((ball) =>
    ball.kind === "cue"
      ? { ...ball, pocketed: ball.pocketed || cuePotted || pocketed.has(ball.id) }
      : { ...ball, pocketed: ball.pocketed || pocketed.has(ball.id) },
  );
}

function colorSpotPosition(
  kind: string,
  balls: readonly BilliardsBall[],
  ignoredBallId: string,
): { x: number; y: number } | null {
  const table = tableSpecFor("snooker");
  const own = table.spots.find((spot) => spot.id === kind);
  if (!own) return null;
  const radius = table.ballDiameter / 2;
  const occupied = (x: number, y: number) =>
    balls.some(
      (ball) =>
        ball.id !== ignoredBallId &&
        !ball.pocketed &&
        Math.hypot(x - ball.x, y - ball.y) < table.ballDiameter - 1e-9,
    );
  if (!occupied(own.x, own.y)) return own;

  // If a color's own spot is occupied, the rules use the highest-value free
  // spot. The final fallback searches along the table's longitudinal centre
  // line, which is the prescribed direction when all spots are occupied.
  const byValue = [...SNOOKER_COLORS_ASCENDING].reverse();
  for (const candidate of byValue) {
    const spot = table.spots.find(({ id }) => id === candidate);
    if (spot && !occupied(spot.x, spot.y)) return spot;
  }
  for (let step = 1; step < 500; step += 1) {
    const x = Math.min(table.width - radius, own.x + (step * radius) / 2);
    if (!occupied(x, own.y)) return { x, y: own.y };
    if (x >= table.width - radius) break;
  }
  for (let step = 1; step < 500; step += 1) {
    const x = Math.max(radius, own.x - (step * radius) / 2);
    if (!occupied(x, own.y)) return { x, y: own.y };
    if (x <= radius) break;
  }
  return null;
}

function respotColors(
  balls: readonly BilliardsBall[],
  ids: readonly string[],
): readonly BilliardsBall[] {
  const colorIds = [...new Set(ids)].filter((id) => {
    const ball = balls.find(({ id: candidate }) => candidate === id);
    return ball !== undefined && isSnookerColor(ball.kind);
  });
  // Higher value colors are placed first if one shot somehow pots several.
  colorIds.sort((a, b) => {
    const first = balls.find(({ id }) => id === a)?.value ?? 0;
    const second = balls.find(({ id }) => id === b)?.value ?? 0;
    return second - first;
  });
  let next = balls;
  for (const id of colorIds) {
    const ball = next.find(({ id: candidate }) => candidate === id);
    if (!ball || !isSnookerColor(ball.kind)) continue;
    const position = colorSpotPosition(ball.kind, next, ball.id);
    if (!position) continue;
    next = next.map((candidate) =>
      candidate.id === id
        ? { ...candidate, pocketed: false, x: position.x, y: position.y }
        : candidate,
    );
  }
  return next;
}

function redsRemaining(balls: readonly BilliardsBall[]): number {
  return balls.filter((ball) => ball.kind === "red" && !ball.pocketed).length;
}

function colorsRemaining(balls: readonly BilliardsBall[]): number {
  return balls.filter((ball) => isSnookerColor(ball.kind) && !ball.pocketed).length;
}

function nextColorAfter(color: SnookerOn): SnookerOn | null {
  const index = SNOOKER_COLORS_ASCENDING.findIndex((candidate) => candidate === color);
  if (index < 0) return null;
  return SNOOKER_COLORS_ASCENDING[index + 1] ?? null;
}

function requireNomination(
  state: Readonly<BilliardsMatchState>,
  nominatedColor: string | null,
): void {
  if (state.snookerOn === "color" && nominatedColor === null) {
    throw new GameRuleError("COLOR_NOMINATION_REQUIRED");
  }
}

function finalBlackResolution(
  state: Readonly<BilliardsMatchState>,
  balls: readonly BilliardsBall[],
  players: BilliardsMatchState["players"],
  actorSeatId: SeatId,
  simulation: Readonly<BilliardsSimulationResult>,
  foulCode: string | null,
  points: number,
  random: GameRandomV1,
): AdjudicatedBilliardsShot {
  const [firstPlayer, secondPlayer] = requireCompetitivePlayers(players);
  const tie = firstPlayer.score === secondPlayer.score;
  const base: BilliardsMatchState = {
    ...state,
    balls,
    breakShot: false,
    lastShot: {
      foulCode,
      points,
      pottedBallIds: [...simulation.pocketedBallIds],
      seatId: actorSeatId,
    },
    players,
    shotNumber: state.shotNumber + 1,
  };
  if (tie) {
    const blackId = balls.find((ball) => ball.kind === "black")?.id;
    const withCue = markCueBallInHand(balls);
    const respotted = blackId ? respotColors(withCue, [blackId]) : withCue;
    const chooserSeatId = random.pick(
      state.seatIds,
      `billiards.snooker.deciding-black.${state.shotNumber + 1}.chooser`,
    );
    return {
      foulCode,
      points,
      state: {
        ...base,
        activeSeatId: chooserSeatId,
        ballInHandZone: null,
        balls: respotted,
        decidingBlack: true,
        outcome: null,
        pendingDecision: {
          chooserSeatId,
          choices: ["play-self", "defer"],
          type: "deciding-black-choice",
        },
        phase: "decision",
        snookerOn: "black",
      },
    };
  }
  const winnerSeatId =
    firstPlayer.score > secondPlayer.score ? firstPlayer.seatId : secondPlayer.seatId;
  const ended = endState(
    { ...base, decidingBlack: false, phase: "aiming", snookerOn: "black" },
    winnerSeatId,
    "final-black",
  );
  return { foulCode, points, state: ended };
}

/** Apply one authoritative snooker shot, including scoring and respotting. */
export function adjudicateSnookerShot(
  input: Readonly<ShotAdjudicationInput> & { readonly random: GameRandomV1 },
): AdjudicatedBilliardsShot {
  const { actorSeatId, simulation, state, shot } = input;
  requireNomination(state, shot.nominatedColor);
  const opponentSeatId = otherSeat(state, actorSeatId);
  const potted = newlyPottedBalls(state, simulation.pocketedBallIds);
  const objectPotted = potted.filter((ball) => ball.kind !== "cue");
  const cuePotted = simulation.cueBallPotted || potted.some((ball) => ball.kind === "cue");
  const firstBalls = simulation.firstContactBallIds
    .map((id) => ballById(state.balls, id))
    .filter((ball): ball is BilliardsBall => ball !== undefined);
  const jumpedBalls = simulation.jumpedBallIds
    .map((id) => ballById(state.balls, id))
    .filter((ball): ball is BilliardsBall => ball !== undefined);
  const on = state.snookerOn ?? "red";
  const targetKind = on === "red" ? "red" : on === "color" ? shot.nominatedColor : on;
  if (targetKind === null) throw new GameRuleError("COLOR_NOMINATION_REQUIRED");
  const targetValue = targetKind === "red" ? 1 : SNOOKER_COLOR_VALUES[targetKind];
  const wrongFirst = firstBalls.length === 0 || firstBalls.some((ball) => ball.kind !== targetKind);
  const wrongPotted = objectPotted.some((ball) => ball.kind !== targetKind);
  const noContact = simulation.firstContactBallIds.length === 0;

  let foulCode: string | null = null;
  if (noContact) foulCode = "NO_BALL_CONTACT";
  else if (wrongFirst) foulCode = "WRONG_FIRST_CONTACT";
  else if (wrongPotted) foulCode = "WRONG_BALL_POTTED";
  if (cuePotted) foulCode = "CUE_BALL_POTTED";
  if (simulation.jumpedBallIds.length > 0) foulCode = "JUMP_SHOT";

  let balls = withPottedFlags(simulation.balls, simulation.pocketedBallIds, cuePotted);
  const redsAfter = redsRemaining(balls);
  const pottedRedCount = objectPotted.filter((ball) => ball.kind === "red").length;
  const targetPotted = objectPotted.some((ball) => ball.kind === targetKind);
  const points =
    foulCode === null
      ? on === "red"
        ? pottedRedCount
        : targetPotted
          ? targetValue
          : 0
      : Math.max(
          4,
          targetValue,
          ...firstBalls.map((ball) => ball.value),
          ...jumpedBalls.map((ball) => ball.value),
          ...objectPotted.map((ball) => ball.value),
        );

  // Colors are always returned to their spots while reds remain on the table;
  // during the clearance phase only the target color stays down on a legal pot.
  const respotIds =
    redsAfter > 0 || on === "red" || on === "color"
      ? objectPotted.filter((ball) => isSnookerColor(ball.kind)).map((ball) => ball.id)
      : foulCode !== null
        ? objectPotted.filter((ball) => isSnookerColor(ball.kind)).map((ball) => ball.id)
        : objectPotted
            .filter((ball) => isSnookerColor(ball.kind) && ball.kind !== targetKind)
            .map((ball) => ball.id);
  if (respotIds.length > 0) balls = respotColors(balls, respotIds);

  let players = state.players;
  if (foulCode === null) {
    players = addScore(players, actorSeatId, points);
  } else {
    players = addScore(players, opponentSeatId, points);
  }
  players = players as BilliardsMatchState["players"];

  const finalBlack = on === "black" && redsAfter === 0 && colorsRemaining(balls) <= 1;
  if (finalBlack && (foulCode !== null || targetPotted)) {
    return finalBlackResolution(
      state,
      balls,
      players,
      actorSeatId,
      simulation,
      foulCode,
      points,
      input.random,
    );
  }

  let nextOn: SnookerOn;
  let continueTurn = false;
  if (foulCode !== null) {
    if (on === "red") nextOn = redsAfter > 0 ? "red" : "color";
    else if (on === "color") nextOn = redsAfter > 0 ? "red" : "yellow";
    else nextOn = on;
  } else if (on === "red") {
    if (pottedRedCount > 0) {
      nextOn = "color";
      continueTurn = true;
    } else {
      nextOn = redsAfter > 0 ? "red" : "color";
    }
  } else if (on === "color") {
    nextOn = redsAfter > 0 ? "red" : "yellow";
    continueTurn = targetPotted;
  } else {
    nextOn = targetPotted ? (nextColorAfter(on) ?? "black") : on;
    continueTurn = targetPotted;
  }

  const ballInHand = cuePotted;
  const nextSeatId = continueTurn && foulCode === null ? actorSeatId : opponentSeatId;
  const nextBalls = ballInHand ? markCueBallInHand(balls) : balls;
  return {
    foulCode,
    points,
    state: {
      ...state,
      activeSeatId: nextSeatId,
      ballInHandZone: ballInHand ? "d" : null,
      balls: nextBalls,
      breakShot: false,
      lastShot: {
        foulCode,
        points,
        pottedBallIds: [...simulation.pocketedBallIds],
        seatId: actorSeatId,
      },
      outcome: null,
      phase: ballInHand ? "ball_in_hand" : "aiming",
      players,
      shotNumber: state.shotNumber + 1,
      snookerOn: nextOn,
    },
  };
}

export function resolveSnookerDecidingBlackChoice(
  state: Readonly<BilliardsMatchState>,
  actorSeatId: SeatId,
  choice: BilliardsDecidingBlackChoice,
): BilliardsMatchState {
  const pending = state.pendingDecision;
  if (
    state.settings.mode !== "snooker" ||
    !state.decidingBlack ||
    state.phase !== "decision" ||
    pending?.type !== "deciding-black-choice"
  ) {
    throw new GameRuleError("NO_DECIDING_BLACK_DECISION_PENDING");
  }
  if (pending.chooserSeatId !== actorSeatId) throw new GameRuleError("NOT_YOUR_TURN");
  if (!pending.choices.includes(choice)) {
    throw new GameRuleError("DECIDING_BLACK_CHOICE_NOT_AVAILABLE");
  }
  return {
    ...state,
    activeSeatId: choice === "play-self" ? actorSeatId : otherSeat(state, actorSeatId),
    ballInHandZone: "d",
    pendingDecision: null,
    phase: "ball_in_hand",
  };
}

File: games/billiards/native/src/rules/snooker.rs
use super::common::{
    SNOOKER_COLORS_ASCENDING, add_score, ball_by_id, end_state, is_snooker_color,
    mark_cue_ball_in_hand, newly_potted_balls, other_seat, require_competitive_players,
    snooker_color_for_on, snooker_color_id, snooker_color_kind, snooker_color_value,
};
use super::error::{RuleError, RuleResult};
use super::types::{
    AdjudicatedBilliardsShot, BallInHandZone, BilliardsBall, BilliardsBallKind,
    BilliardsDecidingBlackChoice, BilliardsEndReason, BilliardsLastShot, BilliardsMatchState,
    BilliardsMode, BilliardsPendingDecision, BilliardsPhase, ShotAdjudicationInput, SnookerColor,
    SnookerOn,
};
use crate::geometry::table_spec;
use std::collections::HashSet;

fn with_potted_flags(
    balls: &[BilliardsBall],
    pocketed_ids: &[String],
    cue_potted: bool,
) -> Vec<BilliardsBall> {
    let pocketed: HashSet<_> = pocketed_ids.iter().map(String::as_str).collect();
    balls
        .iter()
        .cloned()
        .map(|mut ball| {
            ball.pocketed = ball.pocketed
                || pocketed.contains(ball.id.as_str())
                || ball.kind == BilliardsBallKind::Cue && cue_potted;
            ball
        })
        .collect()
}

fn color_from_kind(kind: BilliardsBallKind) -> Option<SnookerColor> {
    match kind {
        BilliardsBallKind::Yellow => Some(SnookerColor::Yellow),
        BilliardsBallKind::Green => Some(SnookerColor::Green),
        BilliardsBallKind::Brown => Some(SnookerColor::Brown),
        BilliardsBallKind::Blue => Some(SnookerColor::Blue),
        BilliardsBallKind::Pink => Some(SnookerColor::Pink),
        BilliardsBallKind::Black => Some(SnookerColor::Black),
        _ => None,
    }
}

fn color_spot_position(
    color: SnookerColor,
    balls: &[BilliardsBall],
    ignored_ball_id: &str,
) -> Option<(f64, f64)> {
    let table = table_spec(BilliardsMode::Snooker);
    let own = table
        .spots
        .iter()
        .find(|spot| spot.id == snooker_color_id(color))
        .copied()?;
    let radius = table.ball_diameter / 2.0;
    let occupied = |x: f64, y: f64| {
        balls.iter().any(|ball| {
            ball.id != ignored_ball_id
                && !ball.pocketed
                && (x - ball.x).hypot(y - ball.y) < table.ball_diameter - 1e-9
        })
    };
    if !occupied(own.x, own.y) {
        return Some((own.x, own.y));
    }

    for candidate in SNOOKER_COLORS_ASCENDING.iter().rev().copied() {
        if let Some(spot) = table
            .spots
            .iter()
            .find(|spot| spot.id == snooker_color_id(candidate))
            && !occupied(spot.x, spot.y)
        {
            return Some((spot.x, spot.y));
        }
    }

    for step in 1..500 {
        let x = (own.x + step as f64 * radius / 2.0).min(table.width - radius);
        if !occupied(x, own.y) {
            return Some((x, own.y));
        }
        if x >= table.width - radius {
            break;
        }
    }
    for step in 1..500 {
        let x = (own.x - step as f64 * radius / 2.0).max(radius);
        if !occupied(x, own.y) {
            return Some((x, own.y));
        }
        if x <= radius {
            break;
        }
    }
    None
}

fn respot_colors(balls: &[BilliardsBall], ids: &[String]) -> Vec<BilliardsBall> {
    let mut seen = HashSet::new();
    let mut color_ids: Vec<_> = ids
        .iter()
        .filter(|id| seen.insert(id.as_str()))
        .filter(|id| ball_by_id(balls, id).is_some_and(|ball| is_snooker_color(ball.kind)))
        .cloned()
        .collect();
    color_ids.sort_by(|first, second| {
        let first_value = ball_by_id(balls, first).map_or(0, |ball| ball.value);
        let second_value = ball_by_id(balls, second).map_or(0, |ball| ball.value);
        second_value.cmp(&first_value)
    });

    let mut next = balls.to_vec();
    for id in color_ids {
        let Some(ball) = ball_by_id(&next, &id) else {
            continue;
        };
        let Some(color) = color_from_kind(ball.kind) else {
            continue;
        };
        let Some((x, y)) = color_spot_position(color, &next, &id) else {
            continue;
        };
        for candidate in &mut next {
            if candidate.id == id {
                candidate.pocketed = false;
                candidate.x = x;
                candidate.y = y;
            }
        }
    }
    next
}

fn reds_remaining(balls: &[BilliardsBall]) -> usize {
    balls
        .iter()
        .filter(|ball| ball.kind == BilliardsBallKind::Red && !ball.pocketed)
        .count()
}

fn colors_remaining(balls: &[BilliardsBall]) -> usize {
    balls
        .iter()
        .filter(|ball| is_snooker_color(ball.kind) && !ball.pocketed)
        .count()
}

fn next_color_after(on: SnookerOn) -> Option<SnookerOn> {
    match on {
        SnookerOn::Yellow => Some(SnookerOn::Green),
        SnookerOn::Green => Some(SnookerOn::Brown),
        SnookerOn::Brown => Some(SnookerOn::Blue),
        SnookerOn::Blue => Some(SnookerOn::Pink),
        SnookerOn::Pink => Some(SnookerOn::Black),
        SnookerOn::Black | SnookerOn::Red | SnookerOn::Color => None,
    }
}

fn require_nomination(input: &ShotAdjudicationInput) -> RuleResult<()> {
    if input.state.snooker_on == Some(SnookerOn::Color) && input.shot.nominated_color.is_none() {
        return Err(RuleError::rule(
            "COLOR_NOMINATION_REQUIRED",
            "A color must be nominated when color is on",
        ));
    }
    Ok(())
}

fn final_black_resolution(
    input: &ShotAdjudicationInput,
    balls: Vec<BilliardsBall>,
    players: Vec<super::types::BilliardsPlayerState>,
    foul_code: Option<&str>,
    points: u32,
    chooser_index: usize,
) -> RuleResult<AdjudicatedBilliardsShot> {
    let (tie, winner_seat_id) = {
        let (first_player, second_player) = require_competitive_players(&players)?;
        (
            first_player.score == second_player.score,
            if first_player.score > second_player.score {
                first_player.seat_id.clone()
            } else {
                second_player.seat_id.clone()
            },
        )
    };
    let mut base = input.state.clone();
    base.balls = balls;
    base.break_shot = false;
    base.last_shot = Some(BilliardsLastShot {
        foul_code: foul_code.map(str::to_owned),
        points,
        potted_ball_ids: input.simulation.pocketed_ball_ids.clone(),
        seat_id: input.actor_seat_id.clone(),
    });
    base.players = players;
    base.shot_number = input.state.shot_number.saturating_add(1);

    if tie {
        let chooser_seat_id = input
            .state
            .seat_ids
            .get(chooser_index)
            .cloned()
            .ok_or_else(|| {
                RuleError::invalid(
                    "INVALID_DECIDING_BLACK_CHOOSER_INDEX",
                    "The deciding-black chooser index is outside the seat list",
                )
            })?;
        let black_id = base
            .balls
            .iter()
            .find(|ball| ball.kind == BilliardsBallKind::Black)
            .map(|ball| ball.id.clone());
        let with_cue = mark_cue_ball_in_hand(&base.balls);
        base.balls = black_id
            .map(|id| respot_colors(&with_cue, &[id]))
            .unwrap_or(with_cue);
        base.active_seat_id = Some(chooser_seat_id.clone());
        base.ball_in_hand_zone = None;
        base.deciding_black = true;
        base.outcome = None;
        base.pending_decision = Some(BilliardsPendingDecision::DecidingBlackChoice {
            chooser_seat_id,
            choices: vec![
                BilliardsDecidingBlackChoice::PlaySelf,
                BilliardsDecidingBlackChoice::Defer,
            ],
        });
        base.phase = BilliardsPhase::Decision;
        base.snooker_on = Some(SnookerOn::Black);
        return Ok(AdjudicatedBilliardsShot {
            foul_code: foul_code.map(str::to_owned),
            points,
            state: base,
        });
    }

    base.deciding_black = false;
    base.phase = BilliardsPhase::Aiming;
    base.snooker_on = Some(SnookerOn::Black);
    let ended = end_state(&base, winner_seat_id, BilliardsEndReason::FinalBlack);
    Ok(AdjudicatedBilliardsShot {
        foul_code: foul_code.map(str::to_owned),
        points,
        state: ended,
    })
}

/// Applies one authoritative snooker shot. `chooser_index` is only consulted
/// when the shot creates a tied, respotted-black decision.
pub fn adjudicate_snooker_shot(
    input: &ShotAdjudicationInput,
    chooser_index: usize,
) -> RuleResult<AdjudicatedBilliardsShot> {
    require_nomination(input)?;
    let state = &input.state;
    let actor_seat_id = input.actor_seat_id.as_str();
    let opponent_seat_id = other_seat(state, actor_seat_id)?;
    let potted = newly_potted_balls(state, &input.simulation.pocketed_ball_ids);
    let object_potted: Vec<_> = potted
        .iter()
        .filter(|ball| ball.kind != BilliardsBallKind::Cue)
        .cloned()
        .collect();
    let cue_potted = input.simulation.cue_ball_potted
        || potted
            .iter()
            .any(|ball| ball.kind == BilliardsBallKind::Cue);
    let first_balls: Vec<_> = input
        .simulation
        .first_contact_ball_ids
        .iter()
        .filter_map(|id| ball_by_id(&state.balls, id))
        .collect();
    let jumped_balls: Vec<_> = input
        .simulation
        .jumped_ball_ids
        .iter()
        .filter_map(|id| ball_by_id(&state.balls, id))
        .collect();

    let on = state.snooker_on.unwrap_or(SnookerOn::Red);
    let target_kind = match on {
        SnookerOn::Red => BilliardsBallKind::Red,
        SnookerOn::Color => {
            let nominated = input.shot.nominated_color.ok_or_else(|| {
                RuleError::rule(
                    "COLOR_NOMINATION_REQUIRED",
                    "A color must be nominated when color is on",
                )
            })?;
            snooker_color_kind(nominated)
        }
        ordered => snooker_color_kind(snooker_color_for_on(ordered).ok_or_else(|| {
            RuleError::invalid(
                "INVALID_SNOOKER_ON",
                "The current snooker target cannot be resolved",
            )
        })?),
    };
    let target_value = if target_kind == BilliardsBallKind::Red {
        1
    } else {
        color_from_kind(target_kind).map_or(0, snooker_color_value)
    };
    let wrong_first =
        first_balls.is_empty() || first_balls.iter().any(|ball| ball.kind != target_kind);
    let wrong_potted = object_potted.iter().any(|ball| ball.kind != target_kind);
    let no_contact = input.simulation.first_contact_ball_ids.is_empty();

    let mut foul_code = if no_contact {
        Some("NO_BALL_CONTACT")
    } else if wrong_first {
        Some("WRONG_FIRST_CONTACT")
    } else if wrong_potted {
        Some("WRONG_BALL_POTTED")
    } else {
        None
    };
    if cue_potted {
        foul_code = Some("CUE_BALL_POTTED");
    }
    if !input.simulation.jumped_ball_ids.is_empty() {
        foul_code = Some("JUMP_SHOT");
    }

    let mut balls = with_potted_flags(
        &input.simulation.balls,
        &input.simulation.pocketed_ball_ids,
        cue_potted,
    );
    let reds_after = reds_remaining(&balls);
    let potted_red_count = object_potted
        .iter()
        .filter(|ball| ball.kind == BilliardsBallKind::Red)
        .count() as u32;
    let target_potted = object_potted.iter().any(|ball| ball.kind == target_kind);
    let points = if foul_code.is_none() {
        if on == SnookerOn::Red {
            potted_red_count
        } else if target_potted {
            target_value
        } else {
            0
        }
    } else {
        let mut penalty = 4_u32.max(target_value);
        for ball in first_balls
            .iter()
            .copied()
            .chain(jumped_balls.iter().copied())
            .chain(object_potted.iter())
        {
            penalty = penalty.max(u32::from(ball.value));
        }
        penalty
    };

    let respot_ids: Vec<_> = if reds_after > 0 || on == SnookerOn::Red || on == SnookerOn::Color {
        object_potted
            .iter()
            .filter(|ball| is_snooker_color(ball.kind))
            .map(|ball| ball.id.clone())
            .collect()
    } else if foul_code.is_some() {
        object_potted
            .iter()
            .filter(|ball| is_snooker_color(ball.kind))
            .map(|ball| ball.id.clone())
            .collect()
    } else {
        object_potted
            .iter()
            .filter(|ball| is_snooker_color(ball.kind) && ball.kind != target_kind)
            .map(|ball| ball.id.clone())
            .collect()
    };
    if !respot_ids.is_empty() {
        balls = respot_colors(&balls, &respot_ids);
    }

    let players = if foul_code.is_none() {
        add_score(&state.players, actor_seat_id, points)
    } else {
        add_score(&state.players, &opponent_seat_id, points)
    };
    let final_black = on == SnookerOn::Black && reds_after == 0 && colors_remaining(&balls) <= 1;
    if final_black && (foul_code.is_some() || target_potted) {
        return final_black_resolution(input, balls, players, foul_code, points, chooser_index);
    }

    let (next_on, continue_turn) = if foul_code.is_some() {
        match on {
            // If the final red was removed during a foul, no optional color is
            // due: the incoming player starts the ordered clearance at yellow.
            SnookerOn::Red => (
                if reds_after > 0 {
                    SnookerOn::Red
                } else {
                    SnookerOn::Yellow
                },
                false,
            ),
            SnookerOn::Color => (
                if reds_after > 0 {
                    SnookerOn::Red
                } else {
                    SnookerOn::Yellow
                },
                false,
            ),
            ordered => (ordered, false),
        }
    } else {
        match on {
            SnookerOn::Red if potted_red_count > 0 => (SnookerOn::Color, true),
            SnookerOn::Red => (
                if reds_after > 0 {
                    SnookerOn::Red
                } else {
                    SnookerOn::Color
                },
                false,
            ),
            SnookerOn::Color => (
                if reds_after > 0 {
                    SnookerOn::Red
                } else {
                    SnookerOn::Yellow
                },
                target_potted,
            ),
            ordered => (
                if target_potted {
                    next_color_after(ordered).unwrap_or(SnookerOn::Black)
                } else {
                    ordered
                },
                target_potted,
            ),
        }
    };

    let ball_in_hand = cue_potted;
    let next_seat_id = if continue_turn && foul_code.is_none() {
        input.actor_seat_id.clone()
    } else {
        opponent_seat_id
    };
    if ball_in_hand {
        balls = mark_cue_ball_in_hand(&balls);
    }
    let mut next = state.clone();
    next.active_seat_id = Some(next_seat_id);
    next.ball_in_hand_zone = ball_in_hand.then_some(BallInHandZone::D);
    next.balls = balls;
    next.break_shot = false;
    next.last_shot = Some(BilliardsLastShot {
        foul_code: foul_code.map(str::to_owned),
        points,
        potted_ball_ids: input.simulation.pocketed_ball_ids.clone(),
        seat_id: input.actor_seat_id.clone(),
    });
    next.outcome = None;
    next.phase = if ball_in_hand {
        BilliardsPhase::BallInHand
    } else {
        BilliardsPhase::Aiming
    };
    next.players = players;
    next.shot_number = state.shot_number.saturating_add(1);
    next.snooker_on = Some(next_on);
    Ok(AdjudicatedBilliardsShot {
        foul_code: foul_code.map(str::to_owned),
        points,
        state: next,
    })
}

pub fn resolve_snooker_deciding_black_choice(
    state: &BilliardsMatchState,
    actor_seat_id: &str,
    choice: BilliardsDecidingBlackChoice,
) -> RuleResult<BilliardsMatchState> {
    let Some(BilliardsPendingDecision::DecidingBlackChoice {
        chooser_seat_id,
        choices,
    }) = state.pending_decision.as_ref()
    else {
        return Err(RuleError::rule(
            "NO_DECIDING_BLACK_DECISION_PENDING",
            "There is no deciding-black decision to resolve",
        ));
    };
    if state.settings.mode != BilliardsMode::Snooker
        || !state.deciding_black
        || state.phase != BilliardsPhase::Decision
    {
        return Err(RuleError::rule(
            "NO_DECIDING_BLACK_DECISION_PENDING",
            "There is no deciding-black decision to resolve",
        ));
    }
    if chooser_seat_id != actor_seat_id {
        return Err(RuleError::rule(
            "NOT_YOUR_TURN",
            "Only the deciding-black chooser may act",
        ));
    }
    if !choices.contains(&choice) {
        return Err(RuleError::rule(
            "DECIDING_BLACK_CHOICE_NOT_AVAILABLE",
            "The selected deciding-black choice is not available",
        ));
    }

    let active_seat_id = match choice {
        BilliardsDecidingBlackChoice::PlaySelf => actor_seat_id.to_owned(),
        BilliardsDecidingBlackChoice::Defer => other_seat(state, actor_seat_id)?,
    };
    let mut next = state.clone();
    next.active_seat_id = Some(active_seat_id);
    next.ball_in_hand_zone = Some(BallInHandZone::D);
    next.pending_decision = None;
    next.phase = BilliardsPhase::BallInHand;
    Ok(next)
}
